### ISSUE TITLE
Anonymous types

### ISSUE THREAD

**Issue by ozra** [ozra/onyx-lang#49]
For many cases where callbacks are used in some languages, it's common in baselib to require visitor(-like) instances instead.

A Visitor instance has several pros to simple callbacks: "pattern match" types via overloading handler functions for different types, maintaining state in a formal way (as opposed to informal closured vars in closured callbacks).

However, there's no way of passing an anonymous visitor instance, as it is possible with an anonymous lambda.

Thus, it would be handy to be able to.

Perhaps this should be "mindstate" connected to "soft lambdas" (_requiring_ a better name for reasonable association), and then syntax likeness too?

```onyx
some-thing.accept (self) <~ Visitor
   @some-state = 0
   init(@base-something) ->
   visit(node SomeType) -> do-stuff
   visit(node OtherType) -> do-stuff
```

Corresponds to:
```onyx
some-thing.accept MyOneOffVisitor(self)

type MyOneOffVisitor < Visitor
   @some-state = 0
   init(@base-something) ->
   visit(node SomeType) -> do-stuff
   visit(node OtherType) -> do-stuff
```

- The `(arguments)` and `<~` looks like the reverse of "soft lambda"
- The reverse `<` gives homage to the `inherit` symbol in the type declaration
- The rest to the right of the back-arrow simply follows normal type syntax

Instead of a type name, the type-definition is "thrown" directly onto the initializer arguments (empty if not applicable of course) and immediately instantiated.

The use case if of course for when the type will only be used in a call and never any more places, which would make it utterly pointless in naming the type.

**Comment by stugol** [comment]
I like this idea, although I'm not sure where it would be used. In what situations (e.g. in the standard library) do we generally require a visitor?

I think the syntax should be slightly different:
```
some-thing.accept (self) <~
   @some-state = 0
   init(@base-something) ->
   call(node SomeType) -> do-stuff
   call(node OtherType) -> do-stuff
```
Makes sense for a visitor class to support `.()`, so what's this `visit` method for? Also, why the word `Visitor` to the right of the `<~`? It seems redundant.

**Comment by stugol** [comment]
Also, if you're creating types on the fly, doesn't this require duck typing at the receiving end?

**Comment by ozra** [comment]
> I like this idea, although I'm not sure where it would be used. In what situations (e.g. in the standard library) do we generally require a visitor?

Well, that is just one example of use case. In the compiler visitors are like 80% of the code.

> What is meant by "pattern match types via overloading handler functions for different types"?

By implementing for example `visit` (in this particular example) with different type restrictions on the parameters you get a "sort of pattern matching", that was all I meant. Once again, in the compiler, there are around 100 `visit` functions defined for the visitor for each specific pass, differing only in type of node: `transform(node VisibilityModifier) ->`, `transform(node For) ->`, etc.

> I think the syntax should be slightly different:

That "is" the syntax. :-) As mentioned, it is the same as regular type definition except for the args instead of type-name and backwards-curly-arrow. This means that to the right `type-builder-name` and/or `ancestor/base-type-name` comes. And `Reference` is default, so no type means you're inheriting the generic `Reference` in your example.
So, the word `Visitor` is of course a type that's inherited from in the given example (this is the common type used in the compiler as basis for visiting).

You could definitely use your own visitor class and use `.()` instead. Using a named function is good because `visit_before` and `visit_after` etc. becomes more logical, but if such functionalities aren't needed for the graph you're working with, then `.()` could be the more reasonable choice.

> Also, if you're creating types on the fly, doesn't this require duck typing at the receiving end?

The way it would work practically is very simple. "Behind the scenes" a type is defined with a `FooBarTempName4747` (increasing temp-number). This is instantiated at definition site in the code with the construction parameters.

It uses the structural ducktype/nominative cross-over marriage of type resolution as everything else in Onyx.

**Comment by stugol** [comment]
Ah. So if the receiver is expecting a visitor of a specific type, you simply specify a base type when using `<~`?

This is looking pretty useful.

**Comment by ozra** [comment]
Yes, exactly like regular `type`-definition. Also see #56 where I made a suggestion of using it for that use-case.

**Comment by stugol** [comment]
Have we implemented this?

**Comment by ozra** [comment]
Nope, have other things in the pipe-line, but I'll get to it soon - shouldn't be to complicated to get working.

**Comment by ozra** [comment]
Ok, quick re-visit here. Alternative ideas to the head part of def?

```onyx
-- first proposal (A):
foo-with-something (constructor, args) <~
   init(foo, bar) ->

foo-with-something (constructor, args) <~ SomeType
   init(foo, bar) ->

-- alternative (B)
foo-with-something new (constructor, args)
   init(foo, bar) ->

foo-with-something new (constructor, args) < SomeType
   init(foo, bar) ->

-- alternative (C)
foo-with-something (constructor, args) <:
   init(foo, bar) ->

foo-with-something (constructor, args) <: SomeType
   init(foo, bar) ->
```

Other idea?

**Comment by stugol** [comment]
I still like (A). Makes it clear we're declaring a callback. I have a question though. Suppose `foo-with-something` takes arguments?
```
foo-with-something (args) (constructor-args) <~           -- hmm?
```
Suppose the visitor takes no arguments? Do we still need the parens?

Do parens bind more tightly to `<~` than to a function call?
```
fn (args) <~          -- whose args are these?
```
Suggest mandating - or at least encouraging - a lack of spacing between the parens and the arrow: `(args)<~`

**Comment by stugol** [comment]
On that note, strikes me the parens would look ugly as sin on any function call that takes multiple lambdas. My experience with CoffeeScript has borne this out:
```
fn(arg1, callback1, callback2, arg3) ->

fn 3, ~> say "Not gonna...", ~> "...work", 17        -- commas probably don't work here....?

fn 3, (~> say "Bugger..."), (~> "...this!"), 17         -- what are we, lisp!?
```
Occurs to me your new indent-based syntax could help here:
```
fn 3
  ~> say "Pretty..."
  ~> "...good"
  17
```
But for one-liners, it's ugly.

**Comment by ozra** [comment]
The instantiated anonymous type is just that - an instance. If done in arguments to a call, that's what it is.
```onyx
foo(x) -> x.the-action "Hey"

-- create an instance of an anon-type
a = () <~
   the-action(x) -> say "{x} action"

foo a

-- create an instance of an anon-type directly in args
foo () <~
   the-action(x) -> say "{x} action"
```

So, nothing magic, just standard syntax for the rest. The parentheses acts as replacement for type-name and will always be required to not cause any ambiguity.

In coffeescript, iirc, expression sequences are expressed the same as in ecmascript, comma separated. In Onyx you'd simply parenthesize a group of expressions for that, so the comma doesn't have that meaning. In Onyx "multiple assignment" is also syntaxed as destructuring, so the comma doesn't have an ambiguous meaning like in Crystal either. This means that the comma will end an expression. Thus it "simply works". Note though that _lambdas_ use syntax `(params) -> body`, bot `~>` which is for _"soft lambdas"_ (which still needs a better name - corresponds to "blocks" in Crystal [which is a ridiculously ambiguous name]), so:
```onyx
fn(arg1, callback1, callback2, arg3) ->

fn 3, () -> say "Will work...", () -> "...fine", 17
```

Note that parens are always required for lambdas _currently_, we might change that. I go for consistency first. A change like that must first be tried and reasoned about so that it doesn't lock up syntax with hard-to-avoid ambiguities (like JS-expr-sequences and multi-assign has done in Crystal syntax for instance - things rarely come for free - making something more terse can destroy the possibilities for countless other notations that would be more beneficial). 

As for the indent-syntaxes and lambdas, I won't comment on them here, as to keep the issue somewhat on topic.

**Comment by stugol** [comment]
Ah, but what about blocks that have multiple statements?
```
fn 3, () -> say 1; say 2, () -> say 3; say 4, 17       -- ???
```
> Note though that lambdas use syntax `(params) ->` body, not `~>`

But wouldn't `~>` work here just as well?

Also, the question of other arguments `(...) () <~` hasn't been addressed:
```
fn(value, visitor) -> visitor.visit value

fn(3) () <~ visit(n) -> say n        -- probably not valid?
fn (3, () <~ visit(n) -> say n)      -- ugly though, in cases where parens are necessary
```

**Comment by ozra** [comment]
> Ah, but what about blocks that have multiple statements?

Good question, I'll clarify myself: It's all good; the semis don't close the expression block, they just end an expression. The commas close/end an entire expression-block. So your example works fine.

> But wouldn't ~> work here just as well?

No, because there currently can only be one soft-lambda ("block") passed to a function (which incidentally started this issue, from #56)

> Also, the question of other arguments (...) () <~ hasn't been addressed:

I guess I wasn't clear enough in the examples.
```onyx
fn(value, visitor) -> visitor.visit value

fn 3, () <~: visit(n) -> say n        -- one line nesting requires the nest-starter (uses colon here)

fn (3, () <~: visit(n) -> say n)      -- this is a tuple with (3, typ-instance) passed to fn as one arg

-- above mistake is also one more argument against the parenthesized tuple notation...

fn(3, () <~: visit(n) -> say n)       -- this is likely what you intended

fn 3, () <~                           -- indented nest instead of one-lined
   visit(n) -> say n        

x = () <~: visit(n) -> say n           -- once again, clarifying that the anon-type is just instantiated to a regular instance
fn 3, x
```

Alternatively with "new" syntax for comparison:

```onyx
fn(value, visitor) -> visitor.visit value

fn 3, new (): visit(n) -> say n        -- one line nesting requires the nest-starter (uses colon here)

fn (3, new (): visit(n) -> say n)      -- this is a tuple with (3, typ-instance) passed to fn as one arg

-- above mistake is also one more argument against the parenthesized tuple notation...

fn(3, new (): visit(n) -> say n)       -- this is likely what you intended

fn 3, new ()                           -- indented nest instead of one-lined
   visit(n) -> say n        

x = new (): visit(n) -> say n           -- once again, clarifying that the anon-type is just instantiated to a regular instance
fn 3, x
```

The notation does look a bit wicked without a supertype, which case would be most common is hard to say, currently it's definitely visitors (thereby with supertype Visitor, which commonly takes a starting point for visitation, so assume below is within a type (hence self):).
Once, again, just to compare syntax visually:
```onyx
fn(value, visitor) -> visitor.visit value

fn 3, new (self) < Visitor: visit(n) -> say n        -- one line nesting requires the nest-starter (uses colon here)

fn (3, new (self) < Visitor: visit(n) -> say n)      -- this is a tuple with (3, typ-instance) passed to fn as one arg

-- above mistake is also one more argument against the parenthesized tuple notation...

fn(3, new (self) < Visitor: visit(n) -> say n)       -- this is likely what you intended

fn 3, new (self) < Visitor                           -- indented nest instead of one-lined
   visit(n) -> say n        

x = new (self) < Visitor: visit(n) -> say n           -- once again, clarifying that the anon-type is just instantiated to a regular instance
fn 3, x
```

All in all, the syntax is far from clear enough currently :-/ hmmm.

**Comment by stugol** [comment]
I think `<~` is far preferable to `<:` here.

Also, we should allow multiple soft-lambdas in a function call, as already mentioned.
